platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,670 @@
+execute first. This is extremely unlikely, but must be defended against in the Execution layer.
+
+### Hashgraph Module
+
+The Hashgraph module orders events into rounds, and assigns timestamps to events. It is given ordered, persisted
+events from Event Intake. Sometimes when an event is added, it turns out to be the last event that was needed to cause
+an entire "round" of events to come to consensus. When this happens, the Hashgraph module emits a `round`. The round
+includes metadata about the round (the list of judge hashes, the round number, the roster, etc.) along with the events
+that were included in the round, in order, with their consensus-assigned timestamps.
+
+![Hashgraph](hashgraph-module.png)
+
+Rounds are immutable. They are sent "fire and forget" style from the Hashgraph module to other modules that require
+them. Some modules only really need the metadata, or a part of the metadata. Others require the actual round data.
+We will pass the full round info (metadata + events) to all listeners, and they can pull from it what they need.
+
+#### Roster Changes
+
+When the roster changes, the Hashgraph algorithm must be made aware. It needs roster information to be able to come
+to consensus. While several modules must respond to roster changes, it is the Hashgraph module that knows the
+applicable roster for any given round. It is given the roster by the Execution layer when it is instructed to produce
+a new round.
+
+At runtime, there is a service in Execution that exposes to the world API for adding nodes to, or removing nodes from,
+the address book. At some point in time, Execution will decide the time is right to construct a new consensus roster,
+and pass that roster to the consensus system. **This must be deterministic**. Each node in the network must assign the
+same roster to the same rounds. The Execution module is able to say exactly which consensus round should use this new
+roster, since it knows about all the rounds already created, and all the rounds it has authorized the Hashgraph module
+to produce, and can therefore know exactly what the next round to produce is, and can tie the new roster to that round.
+
+The Hashgraph module may then apply that roster to that round **or some round in the future**. When a round is produced
+and sent to Execution, the roster used will be supplied. The reason for this is that 3 rounds are involved when coming
+to consensus -- the round that is coming to consensus and 2 newer rounds. Each of these must have been using a
+previously known roster. So when a new roster is supplied, it is **applied** 3 rounds in the future of the currently
+processing round.
+
+The Hashgraph module also includes a `state` section in the metadata of the round. This is used by the Execution layer
+to persist the Consensus state, for reconnect and for restart. In the state is the roster and the judges for the round,
```

Comment:
Why does the roster need to be in the consensus state, if the execution module is responsible for telling consensus which roster to use? This seems to be left over from the previous "double consensus" strategy.

Reply:
I agree, it doesn't.